### PR TITLE
GH-23, GH-24: Support Queue Arguments

### LINKED ISSUE #23: There is no option to configure the consumer queue
Is there a possibility to set properties like max length or time to live  to the consumer queue?

```
private static Map<String, Object> getTTLMaxArgsQueue() {
        final Map<String, Object> args = new HashMap<>
        args.put("x-message-ttl", Duration.ofSeconds(30).toMillis());
        args.put("x-max-length", 1_000);
        return args;
    }
```

### LINKED ISSUE #24: Allow setting TTL on queues
_From @wwadge on February 2, 2016 10:12_

https://github.com/spring-cloud/spring-cloud-stream/blob/master/spring-cloud-stream-binders/spring-cloud-stream-binder-rabbit/src/main/java/org/springframework/cloud/stream/binder/rabbit/RabbitMessageChannelBinder.java#L453

doesn't provide a way to set a TTL on queues to automatically expire messages; the RMQ key can be set as:

```
args.put("x-message-ttl", ttl);
```

though perhaps this can be cleaned up further to allow any additional map entry to be inserted?

_Copied from original issue: spring-cloud/spring-cloud-stream#302_

### PR DESCRIPTION
**Pull request by garyrussell**
Resolves #23
Resolves #24

- support queue arguments - ttl,expires, max-length, max-length-bytes, max-priority
- also on DLQ (if auto-bound)
- also allow customization of dead letter exchange/routing key
- support DLX/DLQ configuration on the DLQ as well - combined with TTL can be used to route back to primary queue

### CODE REVIEW

File: spring-cloud-stream-binder-rabbit-docs/src/main/asciidoc/overview.adoc [hunk 1]
```diff
@@ -113,6 +125,34 @@ delayedExchange::
   The `x-delayed-type` argument is set to the `exchangeType`.
 +
 Default: `false`.
+dlqDeadLetterExchange::
+  if a DLQ is declared, a DLX to assign to that queue
```
Comment (artembilan) [inline spring-cloud-stream-binder-rabbit-docs/src/main/asciidoc/overview.adoc]: DLQ for DLQ?
If I haven't read a PR commit message, I wouldn't understand what are these all about:

> support DLX/DLQ configuration on the DLQ as well - combined with TTL can be used to route back to primary queue

Any ideas to improve these descriptions?

Thanks

File: spring-cloud-stream-binder-rabbit/src/main/java/org/springframework/cloud/stream/binder/rabbit/RabbitCommonProperties.java [hunk 2]
```diff
@@ -90,4 +177,141 @@ public void setBindingRoutingKey(String routingKey) {
...
+	public void setMaxLengthBytes(Integer maxLengthBytes) {
+		this.maxLengthBytes = maxLengthBytes;
+	}
+
+	public Integer getMaxPriority() {
+		return this.maxPriority;
+	}
+
+	@Range(min = 0, max = 255)
+	public void setMaxPriority(Integer maxPriority) {
```
Comment (artembilan) [inline spring-cloud-stream-binder-rabbit/src/main/java/org/springframework/cloud/stream/binder/rabbit/RabbitCommonProperties.java]: M-m-m. Shouldn't this validation annotation be on the getter ?
At least that is what I see in the app starters...

File: spring-cloud-stream-binder-rabbit/src/main/java/org/springframework/cloud/stream/binder/rabbit/RabbitMessageChannelBinder.java [hunk 3]
```diff
@@ -437,20 +480,33 @@ private RabbitTemplate buildRabbitTemplate(RabbitProducerProperties properties)
 	 * @param baseQueueName   The base name for the queue (including the binder prefix, if any).
 	 * @param routingKey  The routing key for the queue.
 	 * @param autoBindDlq true if the DLQ should be bound.
+	 * @param properties the properties.
 	 */
-	private void autoBindDLQ(final String baseQueueName, String routingKey, String prefix, boolean autoBindDlq) {
+	private void autoBindDLQ(final String baseQueueName, String routingKey, RabbitCommonProperties properties) {
```
Comment (artembilan) [inline spring-cloud-stream-binder-rabbit/src/main/java/org/springframework/cloud/stream/binder/rabbit/RabbitMessageChannelBinder.java]: Time to move to IDEA?
wrongjavadoc

😄

File: spring-cloud-stream-binder-rabbit/src/main/java/org/springframework/cloud/stream/binder/rabbit/RabbitMessageChannelBinder.java [hunk 4]
```diff
@@ -437,20 +480,33 @@ private RabbitTemplate buildRabbitTemplate(RabbitProducerProperties properties)
...
-			final String dlxName = deadLetterExchangeName(prefix);
+			String dlxName = deadLetterExchangeName(properties);
 			final DirectExchange dlx = new DirectExchange(dlxName);
 			declareExchange(dlxName, dlx);
-			declareBinding(dlqName, BindingBuilder.bind(dlq).to(dlx).with(routingKey));
+			if (properties.getDeadLetterRoutingKey() == null) {
+				declareBinding(dlqName, BindingBuilder.bind(dlq).to(dlx).with(routingKey));
+			}
+			else {
+				declareBinding(dlqName, BindingBuilder.bind(dlq).to(dlx).with(properties.getDeadLetterRoutingKey()));
```
Comment (artembilan) [inline spring-cloud-stream-binder-rabbit/src/main/java/org/springframework/cloud/stream/binder/rabbit/RabbitMessageChannelBinder.java]: Looks like duplicate code and `BindingBuilder` can be extracted to the variable and we would have only once `declareBinding()` call.
But, yeah, minor and just IMO...

File: spring-cloud-stream-binder-rabbit-docs/src/main/asciidoc/overview.adoc (hunk 1 above)
Comment (garyrussell) [inline spring-cloud-stream-binder-rabbit-docs/src/main/asciidoc/overview.adoc]: This is a common pattern - lots of questions on SO about this.

`someExchange -> mainqueue -> reject -> DLQ`

configure the DLQ with a TTL (say 30000).
configure the DLQ with a DLX = "someExchange"

Rejected/expired messages get a cumulative `x-death` header and you can determine how many times around the loop we have traveled.

I will add a section to the docs to describe the pattern.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-cloud/spring-cloud-stream-binder-rabbit/commit/19f2e349d0be36f2889de7a36dee6e8961e1ed11

**Pull request closed without merging**